File: backend/src/test/java/ca/metricalsky/winston/entity/VideoEntityTest.java
package ca.metricalsky.winston.entity;

import ca.metricalsky.winston.database.entity.channel.ChannelEntity;
import ca.metricalsky.winston.database.entity.video.VideoEntity;
import ca.metricalsky.winston.test.TestUtils;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;

import java.time.OffsetDateTime;

import static org.assertj.core.api.Assertions.assertThat;

@DataJpaTest
class VideoEntityTest {

    @Autowired
    private TestEntityManager entityManager;

    @Test
    void persistsWithOnlyRequiredFields() {
        var channelEntity = persistChannel();
        var videoEntity = new VideoEntity();
        videoEntity.setId(TestUtils.randomId());
        videoEntity.setChannelId(channelEntity.getId());

        var persistedEntity = entityManager.persistFlushFind(videoEntity);

        assertThat(persistedEntity)
                .hasFieldOrPropertyWithValue("id", videoEntity.getId())
                .hasFieldOrPropertyWithValue("channelId", videoEntity.getChannelId())
                .hasAllNullFieldsOrPropertiesExcept("id", "channelId", "lastFetchedAt");
    }

    @Test
    void persistsWithAllOptionalFields() {
        var channelEntity = persistChannel();
        var videoEntity = new VideoEntity();
        videoEntity.setId(TestUtils.randomId());
        videoEntity.setChannelId(channelEntity.getId());
        videoEntity.setTitle(TestUtils.randomString());
        videoEntity.setDescription(TestUtils.randomString());
        videoEntity.setThumbnailUrl(TestUtils.randomString());
        videoEntity.setPublishedAt(OffsetDateTime.now());

        var persistedEntity = entityManager.persistFlushFind(videoEntity);

        assertThat(persistedEntity)
                .hasFieldOrPropertyWithValue("id", videoEntity.getId())
                .hasFieldOrPropertyWithValue("channelId", videoEntity.getChannelId())
                .hasFieldOrPropertyWithValue("title", videoEntity.getTitle())
                .hasFieldOrPropertyWithValue("description", videoEntity.getDescription())
                .hasFieldOrPropertyWithValue("thumbnailUrl", videoEntity.getThumbnailUrl())
                .hasNoNullFieldsOrPropertiesExcept("comments", "details");
    }

    private ChannelEntity persistChannel() {
        var channelEntity = new ChannelEntity();
        channelEntity.setId(TestUtils.randomId());
        return entityManager.persist(channelEntity);
    }
}

File: database/src/test/kotlin/ca/metricalsky/winston/database/entity/video/VideoEntityTest.kt
package ca.metricalsky.winston.database.entity.video

import ca.metricalsky.winston.database.test.annotation.DatabaseTest
import ca.metricalsky.winston.database.test.faker.DatabaseFaker
import io.kotest.core.spec.style.WordSpec
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager

@DatabaseTest
class VideoEntityTest(
    entityManager: TestEntityManager,
): WordSpec({

    val faker = DatabaseFaker()

    "VideoEntity" should {

        "persist with only required fields" {
            val channel = entityManager.persist(faker.channel().minimalEntity())
            val video = faker.video().minimalEntity(channel)

            entityManager.persistFlushFind(video) shouldNotBeNull {
                id shouldNotBe null
                channelId shouldBe video.channelId
                title shouldBe null
                description shouldBe null
                thumbnailUrl shouldBe null
                publishedAt shouldBe null
                lastFetchedAt shouldNotBe null
                details shouldBe null
                comments shouldBe null
            }
        }

        "persist with all optional fields" {
            val channel = entityManager.persist(faker.channel().minimalEntity())
            val video = faker.video().completeEntity(channel)

            entityManager.persistFlushFind(video) shouldNotBeNull {
                id shouldNotBe null
                channelId shouldBe video.channelId
                title shouldBe video.title
                description shouldBe video.description
                thumbnailUrl shouldBe video.thumbnailUrl
                publishedAt shouldNotBe null
                lastFetchedAt shouldNotBe null
                details shouldBe null
                comments shouldBe null
            }
        }
    }
})
